doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+
+void elektraKeynamePushPart (ElektraKeyname * name, const char * part);
+void elektraKeynamePopPart (ElektraKeyname * name);
+
+ElektraKey * elektraKeyNew (const ElektraKeyname * name);
+ElektraKey * elektraKeyRetain (const ElektraKey * key);
+void elektraKeyRelease (const ElektraKey * key);
+
+const ElektraKeyname * elektraKeyGetName (const ElektraKey * key);
+// Note: creates copy of name
+ElektraErrorCode elektraKeySetName (ElektraKey * key, const ElektraKeyname * name);
+// increments nameLock
+void elektraKeyLockName (ElektraKey * key);
+// decrements nameLock
+void elektraKeyUnlockName (ElektraKey * key);
+bool elektraKeyIsNameLocked (const ElektraKey * key);
+
+const ElektraKeyvalue * elektraKeyGetValue (const ElektraKey * key);
+// Note: creates copy of value
+ElektraErrorCode elektraKeySetValue (ElektraKey * key, const ElektraKeyvalue * value);
+
+/**
+ * @returns a `const ElektraKeyset *` that can safely be cast to `ElektraKeyset *`, iff a non-const `ElektraKey *` was passed as @p key
+ */
+ELEKTRA_CONST ElektraKeyset * elektraKeyGetMeta (const ElektraKey * key);
+
+ElektraErrorCode elektraKeyIsBelow (const ElektraKey * first, const ElektraKey * second);
+ElektraErrorCode elektraKeyCompare (const ElektraKey * first, const ElektraKey * second);
+
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * src, ElektraKeyCopyFlag flags);
+static inline ElektraKey * elektraKeyDup (ElektraKey * key, ElektraKeyCopyFlag flags)
+{
+    ElektraKey * dest = flags & ELEKTRA_KEY_CP_NAME
+        ? elektraKeyNew (elektraKeyGetName (key))
+        : elektraKeyNew (&(ElektraKeyname){.ns = ELEKTRA_NS_CASCADING, .name = "", .size = 0});
+    return elektraKeyCopy (dest, key, (ElektraKeyCopyFlag)(flags & ~ELEKTRA_KEY_CP_NAME));
+}
+
+
+ElektraKeyset * elektraKeysetNew (size_t prealloc);
+ElektraKeyset * elektraKeysetRetain (const ElektraKeyset * ks);
+void elektraKeysetRelease (const ElektraKeyset * ks);
+
+size_t elektraKeysetSize (ElektraKeyset * ks);
+
+ElektraErrorCode elektraKeysetInsert (ElektraKeyset * ks, ElektraKey * key);
+ElektraErrorCode elektraKeysetInsertAll (ElektraKeyset * ks, ElektraKeyset * other);
+static inline ElektraErrorCode elektraKeysetInsertAndRelease (ElektraKeyset * ks, ElektraKey * key)
+{
+    ElektraErrorCode error = elektraKeysetInsert (ks, key);
+    elektraKeyRelease (key);
+    return error;
+}
+
+ElektraKey * elektraKeysetGet (ElektraKeyset * ks, size_t index);
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
+
+void elektraKeysetRemove (ElektraKeyset * ks, size_t index);
+void elektraKeysetRemoveRange (ElektraKeyset * ks, size_t start, size_t end);
+static inline void elektraKeysetClear (ElektraKeyset * ks)
+{
+    elektraRemoveRange (ks, 0, elektraKeysetSize (ks));
+}
+
+size_t elektraKeysetLookup (ElektraKeyset * ks, const ElektraKeyname * name);
+
+size_t elektraKeysetFindHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
```

Comment:
Your favorite function :wink: 

Is the `Find` necessary?

```suggestion
size_t elektraKeysetHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
```

Reply:
> Your favorite function :wink:

The function is simply much more useful than `ksCut` IMO. It also makes a lot of sense with the `*Range` functions.

> Is the `Find` necessary?

I don't like function names without a verb. If I read code that uses APIs that I don't know, I find it much easier to understand, if every function name has a verb that tells me what's happening.

In this case, "find"  IMO indicates:
1. The operation is non-destructive.
2. There is actual work happening. Unlike a "get", a "find" indicates to me that I shouldn't call the function more than I need to and should rather cache results. As a concrete example, calling `elektraKeyGetName` multiple times instead of creating a local variable won't have much real world difference. But calling `elektraKeysetFindHierarchy` (on the same unchanged keyset) over and over again, will waste CPU time.
3. It is similar to a "lookup", but not quite the same.

Reply:
See https://github.com/ElektraInitiative/libelektra/pull/4797/commits/1f2e0b902011e4a37e683749520eb9f258583e76 in #4797